Add TableRowClassifier

kevinyamauchi opened the pull request:
As discussed in #8, this PR adds a classifier for performing classification on rows of a feature table.

Currently, this will accept tables as both pandas DataFrame and a dictionary where the column names are keys and the columns are arrays.

I've added tests that show both table types in `tests/test_table_row_classification.py`. It looks like pytest isn't currently being used, but I think it would be nice to convert the tests in a future PR.

This PR closes #8

kevinyamauchi commented on tests/test_table_row_classification.py:
```diff
@@ -0,0 +1,51 @@
+import numpy as np
+import pandas as pd
```
We might want to add a separate set of dependencies for the test suite. Currently, pandas doesn't need to be a dependency for usage of the library (it's used in type annotation, but we can hide that).

kevinyamauchi replied:
I've just added pandas to the github actions workflow for now. We can add a test optional dependencies in a follow up PR if desired.

kevinyamauchi commented on apoc/_object_classifier.py:
```diff
@@ -180,14 +174,14 @@ def _select_features(self, all_features, features_to_select, labels, ground_trut
         Returns
         -------
-        result:list[vector]
+        result:Dict[str, np.ndarray]
             list of vectors corresponding to the requested features. The vectors are shaped (n+1) for n labels. The
             first element corresponds to background.
         ground_truth: ndimage
             selected elements of provided ground truth where it's not 0
```
From my testing, I think these vectors are actually length n. Is that correct?

kevinyamauchi replied:
(I also need to update the description of the type in the docstring)

haesleinhuepf replied:
Yes, all looks good! The docstring was wrong.

kevinyamauchi commented on apoc/_table_row_classifier.py:
```diff
@@ -0,0 +1,207 @@
+            # if the feature names haven't previously been stored,
+            # store them as a list
+            self._ordered_feature_names = list(feature_table.keys())
+            self.classifier.feature_specification = " ".join(self.ordered_feature_names)
```
My understanding is that the `feature_specification` is a string of the column names, each one separated by a space. Is that correct?

haesleinhuepf replied:
Yes! For backwards-compatibility, also comma-separated lists should work.

kevinyamauchi commented:
Hey, @haesleinhuepf ! I think this is ready for a first round of review.

haesleinhuepf commented on apoc/_table_row_classifier.py:
```diff
@@ -0,0 +1,207 @@
+class TableRowClassifier():
+    def __init__(
+            self,
+            opencl_filename="temp_object_classifier.cl",
```
```suggestion
            opencl_filename="temp_table_row_classifier.cl",
```

haesleinhuepf reviewed (approved):
Hey @kevinyamauchi 

awesome, thanks so much for working on this! As the tests still pass, I'm expecting it's backwards compatible, wonderful! We should add a test that an old classifier file also works with the new classifier. I'm suspecting this will fail because in the old ObjectClassifier.cl files, the classifier will be specified as "ObjectClassifier" and the new classifier may expect "TableRowClassifier" as classifier type. The fix for this can be simple though. 

Also, would you mind adding a demo notebook showing how the API works? That's one of the first things I typically look at :-)

Thanks again and have a great weekend!

Robert

kevinyamauchi commented:
> I'm suspecting this will fail because in the old ObjectClassifier.cl files, the classifier will be specified as "ObjectClassifier" and the new classifier may expect "TableRowClassifier" as classifier type. The fix for this can be simple though.

I  actually think this should work. We pass the classifier name when we instantiate the `TableRowClassifier` in the `ObjectClassifier`. Thus, I think this should be fully backwards compatible.

> Also, would you mind adding a demo notebook showing how the API works? That's one of the first things I typically look at :-)

I can add one before we merge.

kevinyamauchi commented on tests/test_table_row_classification.py:
```diff
@@ -0,0 +1,31 @@
+ground_truth = np.array([1, 1, 2, 2, 1])
+
+
+@pytest.mark.parametrize("feature_table", [feature_table_dict, feature_table_data_frame])
```
@haesleinhuepf , here I used the `parametrize` fixture in pytest to use the same test, but pass in different feature table types. https://docs.pytest.org/en/6.2.x/parametrize.html

kevinyamauchi commented on tests/test_table_row_classification.py:
```diff
@@ -0,0 +1,31 @@
+
+
+@pytest.mark.parametrize("feature_table", [feature_table_dict, feature_table_data_frame])
+def test_table_row_classification(tmpdir, feature_table):
```
@haesleinhuepf , here I used the `tmpdir` to create a temporary directory in which we write the openCL file. This temporary directory automatically gets cleaned up after the test runs. This is especially nice when running tests locally where you don't want your folder to get polluted with old test files. https://docs.pytest.org/en/6.2.x/tmpdir.html

kevinyamauchi commented:
Thanks for the review, @haesleinhuepf ! I have added a notebook demonstrating the `TableRowClassifier` and updated the tests. In case you were curious I made comments on places where I "pytest-ified" the tests. I think this is ready for a final review!

haesleinhuepf reviewed (approved):
Awesome, that all looks fantastic @kevinyamauchi ! 

Thanks so much for working on this!
